docs: deepen threat-modeling guidance and framework landscape (#31)
ossf/security-assessments: guidance/background/threat-modeling/goals.md

> ## Threat Modeling: Goals

Reviewer: Why make this an L2 header?

Author: No clue when/why that change flopped out.

But now I'm annoyed by another problem... Where did the "suggest change" button go...?

<img width="721" height="40" alt="Image" src="https://github.com/user-attachments/assets/95b1d044-ed85-4759-9ce1-7727de3f77f7" />

Reviewer: Replies don't get suggest comment, but you can still open a block with the code style `suggestion`:

```suggestion
# Threat Modeling: Goals
```

**[< Previous: Actions](./actions.md)**

> A *goal*, in threat-modeling terms, is a security property the system is trying to provide — what an attacker would have to violate to "score a point." Equally important is the inverse: the *non-goals,* and the *assumptions* the system relies on. An assumption that turns out to be false is the most common source of security surprise. This page covers all three.

Reviewer 2: ```suggestion
A *goal*, in threat-modeling terms, is a security property the system is trying to provide — what an attacker would have to violate to "score a point." Equally important is the inverse: the *non-goals*, which are things that are considered out of scope. Tied into these are the *assumptions* the system relies on. An assumption that turns out to be false is the most common source of security surprise. This page covers all three.
```

> One of the most important things to do in threat modeling is to understand what an attacker can and cannot do based upon the access they have. In our concept of a “game” this is like the conditions by which the attacker gains points (by violating the goals you have for your system) and the legal moves that the attacker can make toward that end.

Reviewer: Do we still need this sentence given the new one?

Author: I'm fine either way, if you want to suggest a change I'll apply it

Reviewer: Done.

> One of the most important things to do in threat modeling is to understand what an attacker can and cannot do based upon the access they have. In our concept of a “game” this is like the conditions by which the attacker gains points (by violating the goals you have for your system) and the legal moves that the attacker can make toward that end.

Reviewer: ```suggestion
```

(re: [the comment on whether this paragraph is still needed](https://github.com/ossf/security-assessments/pull/31/changes/BASE..318c199549a013c27ce43111888276f29f02df90#r3204692618))

Note that this does not mean you should not implement controls in the following areas! It just means that long held assumptions on the efficacy of these measures should be restated and that they are better used as part of a layered approach to make things harder, instead of infallible controls. They should not be relied on alone to stop a skilled attacker.

### Adapting to modern system boundaries

STRIDE has been around for a long time and parts of it map awkwardly onto modern distributed systems. Two adjustments are worth making explicit:

**Privilege escalation is usually lateral movement.** In a system composed of many isolated actors, the question is rarely "can an attacker become root?" — it's "once an attacker has compromised actor X, what other actors can they reach?" Escalation, in modern terms, is a failure to compartmentalize X from Y. If a compromised microservice can read another microservice's database, you have an escalation problem, regardless of OS-level privilege.

> **Spoofing and escalation often happen at the token, not the password.** Distributed systems authorize most requests with bearer tokens (also called *capabilities*) — an API key, a session token, an OIDC ID token, a SPIFFE SVID. Possessing the token is sufficient; identity is rarely re-checked. So for Eve to act as Bob, she does not need Bob's password — she needs a token a service is willing to accept on Bob's behalf, or she needs to confuse a service into using Bob's token to do something Eve wants. When you reason about Spoofing and Escalation of Privilege in a modern system, reason about how tokens are issued, scoped, transmitted, stored, and revoked — not just about login credentials.

Reviewer 2: ```suggestion
**Spoofing and escalation often happen at the token, not the password.** Distributed systems authorize most requests with bearer tokens (also called *capabilities*) — such as an API key, a session token, an OIDC ID token, or a SPIFFE SVID. Possessing the token is sufficient; identity is rarely re-checked. So for Eve to act as Bob, she does not need Bob's password — she needs a token a service is willing to accept on Bob's behalf, or she needs to confuse a service into using Bob's token to do something Eve wants. When you reason about Spoofing and Escalation of Privilege in a modern system, reason about how tokens are issued, scoped, transmitted, stored, and revoked — not just about login credentials.
```

It turns out that becoming a man-in-the-middle is possible in many scenarios, including wireless attacks in a coffee shop, BGP route hijacking, DNS cache poisoning, etc. While it isn’t trivial for any person to become a man-in-the-middle for a network path between two randomly selected computers, it certainly isn’t unobtainable for a large and important class of attackers.

### Software provided by dependencies is secure so long as we take care when adding them

The realistic stance today is the opposite: assume that some dependency in your tree *will* be compromised at some point in the project's lifetime. Attacks of this kind have become routine — direct dependency takeovers (e.g., maintainer-account compromises on npm/PyPI), typosquatting, malicious updates pushed to widely-used packages, and compromised build systems that inject code into otherwise-clean source. Neglected dependencies that simply never receive security patches are a quieter version of the same problem.

Practical mitigations to consider:

> - **Pin and lock.** Use lockfiles or equivalent to ensure builds reproduce exactly the dependencies you reviewed. Floating version ranges turn an upstream compromise into an immediate downstream compromise.

Reviewer 2: I disagree with this.  I think capturing what you resolved dependencies to in a build is key, but I don't agree that floating version ranges do more harm than good.  

At a minimum, this is a [citation needed] moment.  In the spirit of these pages, we could mention there is a debate about this and give counterpoints.

Author: ```suggestion
- **Pin versions.** Pinning your dependencies versions to a specific commit ID protects in the event of a compromise in the upstream delivery mechanisms. Alternatively, some ecosystems provide immutable releases which will provide a similar protection.
```

> - **Generate and consume SBOMs** so you can answer "which of our releases included version X of dependency Y?" in minutes when an advisory drops.

Reviewer 2: ```suggestion
- **Generate and consume SBOMs** so you can answer "which of our releases included version X of dependency Y?" in minutes when an advisory drops.  However, take care to ensure the SBOMs you produce and consume are accurate.  This means using tools that capture information during the build phase, not using source composition analysis (SCA).
```
